Move template parsing to sexp layer

rgrinberg opened the pull request:
I've removed all my parsing/lexing stuff because I was doing it wrong but the other stuff should start making sense. One thing that needs careful review is the new expander, as it's more complex now and perhaps a lot less efficient.

# TODO

- [ ] Bring back expansion for old templates
- [ ] Add proper escaping of `%`, `}`, `)` to string converter.
- [ ] Implement lexer and parser for old and new format

rgrinberg commented:
@diml the constructor names you've proposed don't really work b/c `Var` is already taken for `item`. How about:

* `Var` -> `Form`
* `Short` -> `Var`
* `Full` -> ???

rgrinberg commented:
@diml btw, should we exclude `:` from atom chars as well? Though I don't think it's necessary to have an escape sequence for this char.

diml commented:
> @diml btw, should we exclude : from atom chars as well? Though I don't think it's necessary to have an escape sequence for this char.

No, I think we can keep it. It's only `%` that's problematic

diml commented:
The pretty printing functions are only for debugging? We can just always pretty print using the new syntax

diml commented:
Ok, I played a bit with the types and came with these ones:

```ocaml
module Template : sig
  type sexp

  type syntax =
    | Percent
    | Percent_braces
    | Dollar_braces (* old jbuild file syntax *)
    | Dollar_parens (* old jbuild file syntax *)

  (* encoding:

     - %var -> Form (loc, Percent, "var", Pnil, rest)
     - %{var} -> Form (loc, Percent_braces, "var", Pnil, rest)
     - %{dep:xxx} ->
         Form (loc, Percent_braces, "dep", Ptext ("xxx", Pnil), rest)
     - %{dep:%root/xxx} ->
         Form (loc, Percent_braces, "dep",
               Pvar (loc, "root", Ptext ("xxx", Pnil)), rest)
  *)
  type parts =
    | Nil
    | Text of string * parts
    | Form of Loc.t * syntax * string * arg list * parts

  and payload =
    | Pnil
    | Ptext of string * payload
    | Pvar  of Loc.t * string * payload
end

...

module Ast : sig
  type sexp = t
  type t =
    | Atom            of Loc.t * Atom.t
    | Quoted_string   of Loc.t * string
    | Template        of Loc.t * Template.parts
    | Quoted_template of Loc.t * Template.parts
    | List            of Loc.t * t list
end
```

They are much closer to what we want to allow. They also have a smaller memory footprint. They don't allow a fully recursive syntax, but I don't believe it's necessary and by keeping things simple will stay away from potential troubles.

diml commented:
BTW, templates should be returned as a single token by the lexer.

diml commented:
Thinking about this again, there is no compelling argument for supporting nested variables. So far, the only request we had was to support `${exe:${SCOPE_ROOT}/foo/bar.exe}`, however that's not really better than `${exe:../../foo/bar.exe}`. Additionally a much better solution to this particular problem would be to allow to make executables visible in all the current project so that we could simply write `(run bar ...)` even though `bar` is not public.

So I suggest that we simply forget about nested variables and only focus on moving the parsing of variables to `src/usexp`.

rgrinberg commented:
@diml btw, your AST doesn't allow for `%(foo)`, is that intentional? I forgot if we are planning to move away from that variant of the syntax.

rgrinberg commented:
Also, curious, why make `parts` include the list structure rather than doing it externally:

```
type part =
  | Text of string
  | Form of Loc.t * syntax * string * payload

type parts = part list
```

Is this just to save a little memory?

rgrinberg commented:
I'm now a little more convinced by removing nesting, because I'm not really a fan of how we're implementing it:

* We're allowing different syntax at every nesting level (this isn't really allowed)
* The redundancy between `parts` and `payload` seems artificial to me.

> The pretty printing functions are only for debugging? We can just always pretty print using the new syntax

Those are also used to reflect back error messages to the user. However, many things would be a lot simpler if we didn't try so hard to preserve the syntax for error messages. I don't think it's so necessary either. Since we already report the location properly, and we can simply include the variable name without stringifying it.

Here's another proposal that doesn't support nesting and simplifies the pretty printing and nesting:

```
module Syntax = struct
  type dune = Brace | Parens
  type jbuild = Short | Long
end

type 'syntax parts =
  | Nil
  | Text of string * 'syntax parts
  | Var of Loc.t * string * string * 'syntax parts * 'syntax

type parts_with_syntax =
  | Dune of Syntax.dune parts
  | Jbuild of Syntax.jbuild parts

type t =
  { quoted: bool
  ; parts: parts_with_syntax
  ; loc: Loc.t
  }
```

* For the var constructors, the first string is they `key` in `%{key:bar}`. When it's `""`, the variable is simply `%{bar}` then.

* We don't allow mixing old and new syntax.

rgrinberg commented:
One thing that's a bit annoying is that we still allow constructing things that aren't valid like `Var (Loc.none, "foo", "bar", Short, Nil)`. Since the first string can only be filled when the syntax is `Long`.

diml commented:
> @diml btw, your AST doesn't allow for %(foo), is that intentional? I forgot if we are planning to move away from that variant of the syntax.

Yes, no point having two syntaxes for variables. `%(foo)` doesn't work as a single atom anyway since `(` and `)` are already used by the s-exp syntax.

> Also, curious, why make parts include the list structure rather than doing it externally:

It also saves one indirection. These values are never used outside of the list and IIRC we don't use generic list functions on them, so it seemed better to me to represent it this way.

> However, I'm not yet convinced by the need for a Template and Quoted_template constructors. Keeping quoted as a flag in Template.t allows us to use this type directly as String_with_vars.t. Otherwise, we'll have to re-assemble the quote flag there anyway. Btw, we aren't storing the loc inside the template anymore? This was also useful for reusing Template.t as String_with_vars.t

I suppose that's fine yh. I might have a gove a bit too far in trying to reduce the number of types...

> Also, I think we should keep the argument to Form in a record if you don't mind. The 5 tuple isn't so convenient.

Same here.

BTW, we can get of the short `%root` syntax. It made sense for nested variables, but on its own it's not really a win.

diml commented:
BTW, we should keep the information of percent/dollar only in Syntax.t. It allows to represent mixed templates that don't exist, but on the other hand it should make the interpretation easier:

```ocaml
match ... with
| Text (s, Nil) -> ...
```

rather than:

```ocaml
match ... with
| Dune (Text (s, Nil)) | Jbuild (Text (s, Nil)) -> ...
```

from the point of view of the expander, we only care about old/new syntax when the variable is not found, which in the new syntax always result in an error. So while this information is necessary, I don't think we should put too much the focus on it.

rgrinberg commented:
Okay, I see. Here's the latest invention then:

```ocaml
type var =
  { loc: Loc.t
  ; key: string
  ; payload: string
  ; parts: parts
  ; syntax: syntax
  }

and parts =
  | Nil of parts
  | Text of string * parts
  | Var of var

type t =
  { quoted: bool
  ; parts: parts
  ; loc: Loc.t
  }
```

Also, previously, I thought that having the `Dune` vs. `Jbuild` syntax distinction available via 1 pattern match would make it easier to implement the different expander behavior. But it isn't necessary anymore, because the expander gets access to the whole variable and can make the decision to error or stringify. This is a bit less efficient as in my previous approach we'd be choosing the behavior once per expansion rather than once per variable, but it's unlikely to be the bottleneck here.

rgrinberg commented:
> It also saves one indirection. These values are never used outside of the list and IIRC we don't use generic list functions on them, so it seemed better to me to represent it this way.

Actually, we do. When doing partial expansions, we rely on `List.rev` on the accumulated list of parts when returning the unexpanded bits. So I can either bring back the lists here or implement a `rev` on parts.

rgrinberg commented:
@diml I've moved some sexp types to `Usexp0`. I think this is better than having some types duplicated in Lexer and Usexp. This problem is more acute b/c now we want to parse the templates entirely in the lexer.

rgrinberg commented:
I've added support for parsing unquoted templates for jbuild and dune files but unfortunately, I've somehow broken locations. The line reports in `misc/run.t` is off by 2 for some reason. @diml any idea what's up with it?

Otherwise, I'll implement the quoted templates and fix the locs later.

Btw, the atom and string parsers can now return templates if they encounter variables. I suppose I should name them more appropriately, unless you have a better idea on how to handle this parsing.

rgrinberg commented:
@diml @emillon this is in a mostly "done" state. It needs #886 to actually work, but it should be testable after that and the removal of parsing in `String_with_vars.t`.

@diml I didn't mean to mangle your formatting, but I'm not sure how you maintain the indentation for .mll files. Do you have a special tuareg config for

```ocaml
{ foo;
  bar;
  baz
}
```

Also, I realize that the accumulation in the atom and the quoted case is needlessly complex. I should be able to unify it if we used a buffer interface for atoms as well. Though there might be a slight performance penalty for that.

diml commented:
For the indentation of mll files I mostly do it by hand :/. I really wish ocp-indent would support it.

BTW, I think the lexer is starting to be complex, I suggest that we split it in two files and have one for the jbuild syntax and one for the new one. Even if there is a bit of code duplication, at least they'll each be simpler to understand. Additionally, it'll be easier to convince ourselves that we are not modifying the old syntax when doing changes to the new one.

I also remembered that we use the s-expression parser to parse the output of `opam config env --sexp`, so it might make sense to keep a trivial lexer for it as well. It should only support the basic syntax and not the fancy things such as variables, block comments, block strings, ...

emillon commented on src/string_with_vars.ml:
```diff
@@ -47,62 +47,69 @@ module Token = struct
-let items_of_string s = of_tokens (Token.tokenise s)
+let items_of_string loc s = of_tokens loc (Token.tokenise s)
+
+let _t : Sexp.Of_sexp.ast -> t = function
```
Is this used? You might want to use another name since it won't be caught by warning 32.

emillon commented on src/string_with_vars.ml:
```diff
@@ -136,74 +143,87 @@ module Partial = struct
-      | None -> loop [] (it :: commit_text acc_text acc) items
+    (string_of_var v) (List.length x)
+
+(* let rec rev_append acc = function
```
You might want to delete this part (if it's not for the remaining TODO parts of course).

emillon commented on src/usexp/lexer.mll:
```diff
@@ -116,38 +227,51 @@ and jbuild_atom acc start = parse
       | Newline -> quoted_string_after_escaped_newline mode lexbuf
       | Other   -> quoted_string                       mode lexbuf
     }
+  | (['$' '%'] as varchar) (['{' '('] as delim) {
```
Is `%(` intended to be a valid combination here, or is this filtered in another place?

emillon commented on src/usexp/lexer.mll:
```diff
@@ -158,50 +282,52 @@ and escape_sequence mode = parse
         | 'r' -> '\r'
         | 'b' -> '\b'
         | 't' -> '\t'
+        | '$' when mode = New_syntax ->
```
Not sure about what are the practices regarding polymorphic comparison here, but you can avoid using it here by matching on `(c, mode)` instead.

emillon commented:
Thanks for the cc - I don't have the whole context, so I just added some comments on the implementation itself.

rgrinberg commented:
Thanks for the review @emillon. #905 is much simpler and ends up addressing most of your comments automatically. Give that PR a read if you have time.
